## Tarea-Practica 2. Noel Alejandro Contreras Acosta

### 1. La fórmula semiempírica de la masa (FSM) En física nuclear, la fórmula de Weizs ̈acker (conocida también como 
### fórmula semiempírica) sirve para evaluar la masa y otras propiedades de un nucleo atómico; y está basada 
### parcialmente en mediciones empíricas. En particular la formula se usa para calcular la  ́energíade enlace 
### nuclear aproximada B, de un nucleo atómico con número atómico Z y número de masa A:
$$
B=a_1A-a_2A^{2/3}-a_3\frac{Z^2}{A^{1/3}}-a_4\frac{(A-2Z)^2}{A}+\frac{a_5}{A^{1/2}}
$$
### donde, en unidades de millones de electron-volts, las constantes son a_1 = 15,8, a_2 = 18,3, a3 = 0,714, a4 = 23,2 y
$$
a_5 =
\begin{cases}
0, & \text{si } A \text{ es impar}, \\
12.0, & \text{si } A \text{ y } Z \text{ son pares (ambos)}, \\
-12.0, & \text{si } A \text{ es par y } Z \text{ impar}.
\end{cases}
$$

#### a) Escribe un programa que tome como entrada los valores de A y Z, e imprima la energía de enlace B para el átomo correspondiente. Usa tu 
#### programa para encontrar la energía de enlace de un átomo con A = 58 y Z = 28 (Hint: La respuesta correcta es alrededor de los 490 MeV).

In [48]:
a1,a2,a3,a4=15.8,18.3,0.714,23.2 #las unidades son MeV
#la función creada aquí pide como argumentos los valores de A y Z, y calcula e imprime el valor de B.
def energia_B(A, Z):
    if A%2==1:
        a5=0
    elif A%2==0 and Z%2==0:
        a5=12
    elif A%2==0 and Z%2==1:
        a5=-12
    B= a1*A-a2*A**(2/3)-(a3*Z*Z)/A**(1/3)-(a4*(A-2*Z)**2)/A+a5/A**(1/2)
    print('La energía de enlace nuclear es',B, 'MeV')

In [3]:
#Prueba para A=58 y Z=28
energia_B(58,28)

#El resultado fue de 497.5620206224374 MeV

La energía de enlace nuclear es 497.5620206224374 MeV


#### b) Modifica el programa del inciso anterior, para escribir una segunda version que  ́imprima no la energía de 
#### enlace total B, sino la energía de unión por nucleón, que es B/A.

In [5]:
def energia_nucleon(A, Z):
    if A%2==1:
        a5=0
    elif A%2==0 and Z%2==0:
        a5=12
    elif A%2==0 and Z%2==1:
        a5=-12
    B= a1*A-a2*A**(2/3)-(a3*Z*Z)/A**(1/3)-(a4*(A-2*Z)**2)/A+a5/A**(1/2)
    x=B/A
    print('La energía de unión por nucleón es ',x, 'MeV')

A=int(input('Ingrese el valor de A: '))
Z=int(input('Ingrese el valor de Z: '))
energia_nucleon(A,Z)

Ingrese el valor de A:  58
Ingrese el valor de Z:  28


La energía de unión por nucleón es  8.578655527973059 MeV


#### c) Escribe una tercera versión del programa para que tome como entrada solo un valor del numero atómico Z y luego pase
#### por todos los valores de A desde A = Z hasta A = 3Z, para encontrar el que tiene la mayor energía de enlace por nucleon.
#### Este es el nucleo más estable con el número atómico dado. Haz que tu programa imprima el valor de A para este nucleo 
#### más estable y el valor de la energía de enlace por nucleón.

In [9]:
def energia_max(Z):
    energias=[] #lista vacia
    for A in range(Z,3*Z+1): #Aquí se recorre los valores de A entre Z y 3Z y,  dependiendo de si A es par, impar (lo mismo para Z), 
        if A%2==1:           #es el valor de B que nos calcula y por ende el de enuc.
            a5=0
        elif A%2==0 and Z%2==0:
            a5=12
        elif A%2==0 and Z%2==1:
            a5=-12
        B= a1*A-a2*A**(2/3)-(a3*Z*Z)/A**(1/3)-(a4*(A-2*Z)**2)/A+a5/A**(1/2)
        enuc=B/A
        energias.append((A, enuc)) #a la lista vacía se le van agregando tuplas de (A, enuc) por cada iteración.
    maxA, maxEnergia = max(energias, key=lambda x: x[1]) #aquí se busca el máximo de cada tupla, tomando para esto a enuc  como el parámetro.
                                                         #además de asociar la A que corresponda a ese valor máximo y se define como maxA
    print(f'El valor de A para este núcleo más estable es {maxA}, y el valor de la energía de enlace por nucleón es {maxEnergia}')


Z=int(input('Ingresa el valor de número atómico Z: '))
energia_max(Z)
        
        
        

Ingresa el valor de número atómico Z:  28


El valor de A para este núcleo más estable es 62, y el valor de la energía de enlace por nucleón es 8.70245768367189


#### d) Finalmente, escribe una cuarta versión del programa que, en lugar de tomar Z como entrada, se ejecute a través de todos los valores 
#### de Z de 1 a 100 e imprima el valor mas estable de A para cada uno. ¿A qué valor de Z se produce la energÍa de enlace máxima por nucleón
#### (La respuesta correcta, en la vida real, es Z = 28, que corresponde al N ́ıquel).

In [11]:
energias=[] #lista vacia
energias_max=[] #lista vacia
for Z in range(1,101):
    
    for A in range(Z,5*Z+1): #Aquí se recorre los valores de A entre Z y 3Z y,  dependiendo de si A es par, impar (lo mismo para Z), 
        if A%2==1:           #es el valor de B que nos calcula y por ende el de enuc.
            a5=0
        elif A%2==0 and Z%2==0:
            a5=12
        elif A%2==0 and Z%2==1:
            a5=-12
        B= a1*A-a2*A**(2/3)-(a3*Z*Z)/A**(1/3)-(a4*(A-2*Z)**2)/A+a5/A**(1/2)
        enuc=B/A
        energias.append((A, enuc)) #a la lista vacía se le van agregando tuplas de (A, enuc) por cada iteración.
    maxA, maxEnergia = max(energias, key=lambda x: x[1]) #aquí se busca el máximo de cada tupla, tomando para esto a enuc  como el parámetro.
                                                         #además de asociar la A que corresponda a ese valor máximo y se define como maxA
    energias_max.append(maxEnergia)
    print('Para un núcleo con número atómico',Z, f'El valor de A que lo hace más estable es {maxA},' 
    f'y el valor de la energía de enlace por nucleón es {maxEnergia}.')
max_abs=max(energias_max)
print('La energía máxima general de enlace nuclear por nucleón es de', max_abs, 'y el primero en que se da es en Z=28.' )

Para un núcleo con número atómico 1 El valor de A que lo hace más estable es 3,y el valor de la energía de enlace por nucleón es 0.36869091831015793.
Para un núcleo con número atómico 2 El valor de A que lo hace más estable es 4,y el valor de la energía de enlace por nucleón es 5.321930578649441.
Para un núcleo con número atómico 3 El valor de A que lo hace más estable es 4,y el valor de la energía de enlace por nucleón es 5.321930578649441.
Para un núcleo con número atómico 4 El valor de A que lo hace más estable es 8,y el valor de la energía de enlace por nucleón es 6.466330085889912.
Para un núcleo con número atómico 5 El valor de A que lo hace más estable es 11,y el valor de la energía de enlace por nucleón es 6.650123444727665.
Para un núcleo con número atómico 6 El valor de A que lo hace más estable es 14,y el valor de la energía de enlace por nucleón es 7.200918138809924.
Para un núcleo con número atómico 7 El valor de A que lo hace más estable es 15,y el valor de la energía de 

## 2. La constante de Madelung
#### En física de la materia condensada, la constante de Madelung da el potencial electrico total que siente un átomo en un  sólido; y depende de las
#### cargas de los otros átomos cercanos y de sus ubicaciones. Por ejemplo, el cristal de cloruro de sodio solido (la sal de mesa), tiene átomos
#### dispuestos en una red cúbica, con átomos de sodio y cloro alternados, teniendo los de sodio una carga positiva +e y los de cloro una 
#### negativa −e, (donde e es la carga del electron). Si etiquetamos cada posición en la red con tres coordenadas enteras (i, j, k), 
#### entonces los atomos de sodio caen en posiciones donde i + j + k es par, y los atomos de cloro en posiciones donde i + j + k es 
#### impar. Consideremos un atomo de sodio en el origen, i.e. i = j = k = 0, y calculemos la constante de Madelung. Si el espaciado 
#### de los atomos en la red es a, entonces la distancia desde el origen al atomo en la posición (i, j, k) es:
$$ \sqrt{(ia)^2+(ja)^2+(ka)^2} = a\sqrt
{i^2+j^2+k^2}$$
#### y el potencial en el origen creado por tal atomo es:
$$
V(i,j,k)= \pm \frac{e}{4\pi\epsilon_oa\sqrt
{i^2+j^2+k^2}}
$$
#### siendo ε0 la permitividad del vacío y el signo de la expresion se toma dependiendo de si i + j + k es par o impar. Así entonces, el potencial 
#### total que siente el átomo de sodio es la suma de esta cantidad sobre todos los demas átomos. Supongamos una caja cúbica alrededor del átomo
#### de sodio en el origen, con L atomos en todas las direcciones, entonces:
$$
V_{\text{total}} = \sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L} V(i,j,k)= \frac{e}{4\pi\epsilon_0a}M
$$
#### donde M es la constante de Madelung (al menos aproximadamente).
#### Técnicamente, la constante de Madelung es el valor de M cuando L → ∞, pero se puede obtener una buena aproximación simplemente usando
#### un valor grande de L. Escribe un programa para calcular e imprimir la constante de Madelung para el cloruro de sodio. Utiliza un 
#### valor de L tan grande  como puedas, sin dejar que tu programa se ejecutar en un tiempo razonable (un minuto o menos).

##### Primero notemos que la expresión 
$$
V_{\text{total}} = \sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L} V(i,j,k)
$$
##### se puede reescribir como
$$
V_{\text{total}} = \sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L}  \pm \frac{e}{4\pi\epsilon_oa\sqrt
{i^2+j^2+k^2}}
$$ 
##### Y se puede ver como:
$$
V_{\text{total}} = \frac{e}{4\pi\epsilon_oa}\sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L}\pm \frac{1}{\sqrt{i^2+j^2+k^2}}
$$
##### De donde vemos que la constante de Madelung es:
$$
M=\sum_{\substack{i,j,k=-L \\ i,j,k \neq 0}}^{L}  \pm \frac{1}{\sqrt
{i^2+j^2+k^2}}
$$

##### Así que el programa debe calcular dicha suma.

In [ ]:
from numpy import sqrt
def constante_Madelung(L):
    M=0
    for i in range(-L,L+1):
        for j in range(-L,L+1):
            for k in range(-L, L+1):
                if i==0 and j==0 and k==0:
                    continue
                m=1/sqrt(i**2+j**2+k**2)
                M+= (-1)**(i+j+k)*m
    return M
L=100
M_a=constante_Madelung(L)
print(f"La constante de Madelung aproximada para NaCl con L={L} es: {M_a}")
    
                
                

## 3. Coeficientes binomiales
#### El coeficiente binomial (nk) es un numero entero igual a:  
$$
  \binom{n}{k}=\frac{n!}{k!(n-k)!}=\frac{n(n-1)(n-2)...(n-k+1)}{1x2x3x...k}
$$
donde $k ≥ 1$, o bien $\binom{n}{0} = 1$ cuando k = 0.

#### a) Utiliza esta fórmula para escribir una función llamada binomial (n,k) (o comotú quieras) que calcule el coeficiente binomial
#### para un n y k dados. Asegurate de que tu funcion devuelva la respuesta en forma de un número entero (no flotante) y proporcione
#### el valor correcto de 1 para el caso en que k = 0.

In [5]:
def binomial(n,k): #La función está deividia en tres casos: cuando k>0, k=0 y k<0.
    if k >=1: 
        numerador=1
        denominador=1
        for i in range(n-k+1,n+1):
            numerador*=i #Aquí se hace la multiplicación que da el valor del numerador dependiendo de n y k.
        for l in range(1,k+1):
            denominador*=l #Aquí se hace la multiplicación que da el valor del denominador dependiendo de n y k.
        coef=numerador/denominador #se obtiene el cociente binomial. 
        return int(coef) #regresa en entero el coeficiente binomial.
    elif k==0:
        x=1
        return x #si k=0, regresa el valor del coeficiente binomial como 1
    elif k<0:
        print('El valor de k no es válido, debe ser mayor o igual que 0. \nInténtalo de nuevo')
        
    
    
    
        
        
    
        
        

In [9]:
binomial(12,8)

495

#### b) Usando tu funcion, escribe un programa que imprima las primeras 20 líneas del “triángulo de Pascal’’. La nesima línea del triangulo
#### de Pascal contiene n + 1 numeros, que son los coeficientes (n 0), (n 1), y así sucesivamente hasta (n n). De tal manera que las 
#### primeras líneas son:
#### 1 1
#### 1 2 1
#### 1 3 3 1
#### 1 4 6 4 1

In [32]:
for n in range(1,21):
    for  k in range(n+1):
        print(end=" ")
        print(int(binomial(n,k)), end=" ")

    print()

 1  1 
 1  2  1 
 1  3  3  1 
 1  4  6  4  1 
 1  5  10  10  5  1 
 1  6  15  20  15  6  1 
 1  7  21  35  35  21  7  1 
 1  8  28  56  70  56  28  8  1 
 1  9  36  84  126  126  84  36  9  1 
 1  10  45  120  210  252  210  120  45  10  1 
 1  11  55  165  330  462  462  330  165  55  11  1 
 1  12  66  220  495  792  924  792  495  220  66  12  1 
 1  13  78  286  715  1287  1716  1716  1287  715  286  78  13  1 
 1  14  91  364  1001  2002  3003  3432  3003  2002  1001  364  91  14  1 
 1  15  105  455  1365  3003  5005  6435  6435  5005  3003  1365  455  105  15  1 
 1  16  120  560  1820  4368  8008  11440  12870  11440  8008  4368  1820  560  120  16  1 
 1  17  136  680  2380  6188  12376  19448  24310  24310  19448  12376  6188  2380  680  136  17  1 
 1  18  153  816  3060  8568  18564  31824  43758  48620  43758  31824  18564  8568  3060  816  153  18  1 
 1  19  171  969  3876  11628  27132  50388  75582  92378  92378  75582  50388  27132  11628  3876  969  171  19  1 
 1  2

#### c) La probabilidad de que para una moneda no sesgada, lanzada n veces, salga aguila k veces es:

$$ p(k|n) = \frac{\binom{n}{k}}
{2^n}
$$ 
#### Escribe un programa para calcular:
#### 1) la probabilidad total de que una moneda lanzada 100 veces, salga aguila exactamente 60 veces y
#### 2) la probabilidad de que salga aguila 60 veces o más.

In [44]:
import math
def binomial(n,k):
  if k == 0:
    return 1
  elif k > n:
    return("n debe ser mayor o igual a k")
  else:
    return((math.factorial(n))/(math.factorial(k)*math.factorial(n-k)))

def probabilidad(n,k):
  return(binomial(n,k)/(2**n))
print(probabilidad(100,60))
for k in range(60, 101):
    print(f"Probabilidad para k = {k} es: {probabilidad(100, k)}")

0.010843866711637987
Probabilidad para k = 60 es: 0.010843866711637987
Probabilidad para k = 61 es: 0.00711073226992655
Probabilidad para k = 62 es: 0.00447287997624412
Probabilidad para k = 63 es: 0.0026979276047186754
Probabilidad para k = 64 es: 0.0015597393964779842
Probabilidad para k = 65 es: 0.0008638556657416528
Probabilidad para k = 66 es: 0.00045810527728724014
Probabilidad para k = 67 es: 0.00023247133474277857
Probabilidad para k = 68 es: 0.00011281697127223077
Probabilidad para k = 69 es: 5.232091421320847e-05
Probabilidad para k = 70 es: 2.3170690580135184e-05
Probabilidad para k = 71 es: 9.790432639493739e-06
Probabilidad para k = 72 es: 3.9433687020183116e-06
Probabilidad para k = 73 es: 1.5125249815960647e-06
Probabilidad para k = 74 es: 5.518672230147804e-07
Probabilidad para k = 75 es: 1.9131397064512386e-07
Probabilidad para k = 76 es: 6.2932227185896e-08
Probabilidad para k = 77 es: 1.9615239642357197e-08
Probabilidad para k = 78 es: 5.78398092018225e-09
Probabilid

### 4. Recursión
#### Una característica muy útil de las funciones que podemos definir es la recursividad, es decir, la capacidad de una funcion de llamarse a sí misma.

#### Por ejemplo, considere la siguiente definicion del factorial n! de un entero positivo n:

$$
n! =
\begin{cases} 
1, & \text{si } n = 1, \\ 
n \times (n-1)!, & \text{si } n > 1.
\end{cases} $$
#### Esto constituye una definición completa del factorial que nos permite calcular el valor de n! para cualquier entero positivo. Podemos emplear 
#### esta definicion directamente para crear una función para calcular factoriales, como esta (en Python):
#### def factorial(n):
####     if n==1:
#### return 1
#### else:
####    return n*factorial(n-1)

#### Nota como si n no es igual a 1, la funcion se llama a sí misma para calcular el factorial de n − 1, de tal manera que al pedir 
#### “print(factorial(5))” el interprete de Python imprimira correctamente la respuesta 120. 
#### a) Anteriormente, en clase encontramos los numeros de Catalán Cn. Si hacemos una pequenñaa modificación, la definició de Cn se puede 
#### reescribir de la siguiente forma:
$$ C_n = \begin{cases}
1, & \text{si } n = 0 \\
\frac{4n − 2}{n + 1}C_{n−1} & \text{si } n > 0
\end{cases}$$

#### Escribe una función, usando recursividad, que calcule Cn y utilízala para calcular e imprimir C_100.

In [36]:
def catalan(n):
    if n==0:
        return 1
    elif n>0:
        return (4*n-2)/(n+1)*catalan(n-1)

print(catalan(100))
    

8.965199470901317e+56


#### b) Euclides demostro que el máximo común divisor g(m, n) de dos enteros no negativos m y n satisface que:
$$
g(m,n) =
\begin{cases} 
m, & \text{si } n = 0, \\ 
g(n, m \bmod n), & \text{si } n > 0.
\end{cases}
$$
#### Usando esta formula escribe una función g(m,n) que emplee recursividad para calcular el maximo común divisor de  m y n. 
#### Usa tu funcion para calcular e imprimir el maximo común divisor de 108 y 192.

In [40]:
def g(m, n):
  if n==0:
    return m
  elif n>0:
    return g(n, m%n)
print(g(108,192))

12
